0000-py-version-support.md

```diff
@@ -0,0 +1,244 @@
+
+Arch
+----
+Arch is a rolling release and is consistantly updating their Python version, so this is not
+of any concern for this policy.
+
+FreeBSD
+-------
+According to this https://wiki.freebsd.org/Python/PortsPolicy FreeBSD currently already aligns
+with this policy. They only support Python versions until they are End of Life so there are
+no concerns with this operating system.
+
+SmartOS
+-------
+SmartOS includes the latest Python version so will not be impacted by this change.
+
+
+What about dependencies?
+------------------------
+Before dropping an End of Life Python version we should have an idea beforehand if Salt's dependencies
+do not support a particular Python version with our testing suite. Currently our tests use the dependencies
+in the requirements directory in the main Salt project. This does of course mean that if we are not testing
+a module we would not have this information beforehand. We are now requiring tests to be added
+for PRs so this will help ensure this is caught for those modules, but there are still some modules that do
+not have tests currently.
+
+For the Sodium Release, I installed the libraries from anything defined with :depends: in the documentation
+and the requirements defined in our requirements directory. Once these were installed I ran
+this tool https://pypi.org/project/checkmyreqs/ against all installed Python libraries. This brought to light
+the following libraries that will not support Python 3, which will impact some Salt components:
+
+ - pycassa (salt/modules/cassandra.py, salt/returners/cassandra_return.py)
+ - vbox (salt/modules/vbox_guest.py, salt/modules/vbox_guest.py,
+         salt/modules/vboxmanage.py, salt/utils/virtualbox.py,
+         salt/cloud/clouds/virtualbox.py)
+
+Both of these projects have not had commits recently added to the project in years. This will need to be documented
+for the Sodium release to make it clear that these modules will not work unless the upstream projects add
+Python 3 support.
+
+Going forward adding more test coverage will help to uncover these projects before we drop a Python version,
+but if we do not have enough test coverage, this exercise should be done any time a Python Version is dropped.
+There might be a conflict between a Salt dependency and its Python version support that will impact many users.
+If this conflict does occur we may need to delay dropping support for the End of Life Python version until that
+dependency is updated to the appropriate Python version.
+
+## Who is Affected
+
+Salt Contributors
+-----------------
+When a Python Version is dropped any contributions will need to ensure they align with the syntax
+of the lowest supported Python version. These contributors can look at Python's Release Notes for
+details around what is added into a release or removed. For example for Python 3.6
+https://docs.Python.org/3/whatsnew/3.6.html outlines some of the changes and removals added to this version.
```

Comment:
Good point. If a module in salt does not work on the minimal supported version, do we have a policy in place to drop those modules?

Reply:
Until Python 4, there shouldn't be any breaking changes. Even if there was, the module should be fixed, not dropped.

Reply:
Even in Python4 there's unlikely to be breaking changes :upside_down_face: